### PR TITLE
possibility to abort long running jobs

### PR DESCRIPTION
**Pull request by david-e**
This patch permits to abort deferred jobs.
It will fix issue #172

### CODE REVIEW

File: arq/jobs.py [hunk 1]
```diff
@@ -6,7 +6,8 @@
 from enum import Enum
 from typing import Any, Callable, Dict, Optional
 
-from .constants import default_queue_name, in_progress_key_prefix, job_key_prefix, result_key_prefix
+from .constants import abort_key_prefix, default_queue_name, in_progress_key_prefix, \
```
Comment (samuelcolvin) [inline arq/jobs.py]: I think this will fail linting, you need to run `make format`.

File: arq/jobs.py [hunk 2]
```diff
@@ -66,6 +67,10 @@ def __init__(
         self._queue_name = _queue_name
         self._deserializer = _deserializer
 
+    async def abort(self, timeout: Optional[float] = None, *, pole_delay: float = 0.5) -> Any:
+        await self._redis.set('%s%s' % (abort_key_prefix, self.job_id), 1, expire=timeout)
```
Comment (samuelcolvin) [inline arq/jobs.py]:
```suggestion
        await self._redis.set(f'{abort_key_prefix}{self.job_id}', b'1', expire=timeout)
```
Please use f-strings or if that's ugly, `.format()`

File: arq/worker.py [hunk 3]
```diff
@@ -275,11 +278,11 @@ def run(self) -> None:
 
             if self.burst:
                 if 0 <= self.max_burst_jobs <= self._jobs_started():
-                    await asyncio.gather(*self.tasks)
+                    await asyncio.gather(*list(self.tasks.values()))
```
Comment (samuelcolvin) [inline arq/worker.py]:
```suggestion
                    await asyncio.gather(*self.tasks.values())
```
I doubt list is required here.

File: arq/worker.py [hunk 4]
```diff
@@ -275,11 +278,11 @@ def run(self) -> None:
...
                     return
                 queued_jobs = await self.pool.zcard(self.queue_name)
                 if queued_jobs == 0:
-                    await asyncio.gather(*self.tasks)
+                    await asyncio.gather(*(self.tasks.values()))
```
Comment (samuelcolvin) [inline arq/worker.py]:
```suggestion
                    await asyncio.gather(*self.tasks.values())
```
These parentheses do nothing.

File: arq/worker.py [hunk 5]
```diff
@@ -295,13 +298,20 @@ def run(self) -> None:
             job_ids = await self.pool.zrangebyscore(
                 self.queue_name, offset=self._queue_read_offset, count=count, max=now
             )
+            self.abort_tasks = [
```
Comment (samuelcolvin) [inline arq/worker.py]: Can we have a switch (default off) for whether we need to abort tasks, so this step can be omitted for those who don't need it. (this is a performance critical part of arq and `keys('...*')` is fairly slow I think.

Does `abort_tasks` need to be `self.abort_tasks`? I don't see why. Also gain, use f-strings.

Also, use a set not a list.

Comment (samuelcolvin) [reply arq/worker.py]: also, see [this](https://redis.io/commands/keys) warning, could we use, scan instead?

File: arq/worker.py [hunk 6]
```diff
@@ -295,13 +298,20 @@ def run(self) -> None:
...
+                k[len(abort_key_prefix):]
+                for k in await self.pool.keys('%s*' % abort_key_prefix)
+            ]
         await self.run_jobs(job_ids)
 
-        for t in self.tasks:
-            if t.done():
-                self.tasks.remove(t)
+        jobs = copy(self.tasks)
```
Comment (samuelcolvin) [inline arq/worker.py]: this can be inlined, probably as `list(jobs.items())`

File: arq/worker.py [hunk 7]
```diff
@@ -432,7 +442,7 @@ def run(self) -> None:
             logger.info('%6.2fs → %s(%s)%s', (start_ms - enqueue_time_ms) / 1000, ref, s, extra)
             # run repr(result) and extra inside try/except as they can raise exceptions
             try:
-                async with async_timeout.timeout(timeout_s):
+                async with async_timeout.timeout(timeout_s or None):
```
Comment (samuelcolvin) [inline arq/worker.py]: why is this required? Is it related to this PR?

File: arq/worker.py [hunk 8]
```diff
@@ -450,9 +460,15 @@ def run(self) -> None:
...
-            elif self.retry_jobs and isinstance(e, asyncio.CancelledError):
-                logger.info('%6.2fs ↻ %s cancelled, will be run again', t, ref)
-                self.jobs_retried += 1
+            elif isinstance(e, asyncio.CancelledError):
+                if job_id in self.abort_tasks:
+                    logger.info('%6.2fs 🛇  %s aborted', t, ref)
+                    result = e
+                    finish = True
+                    self.abort_tasks.remove(job_id)
+                elif self.retry_jobs:
```
Comment (samuelcolvin) [inline arq/worker.py]: technically the `elif, else` logic has changed here.

I think it should be:
```suggestion
            elif isinstance(e, asyncio.CancelledError) and (self.retry_jobs or job_id in self.abort_tasks):
                if job_id in self.abort_tasks:
                    logger.info('%6.2fs 🛇  %s aborted', t, ref)
                    result = e
                    finish = True
                    self.abort_tasks.remove(job_id)
                else:
```

File: arq/worker.py [hunk 9]
```diff
@@ -295,13 +298,20 @@ def run(self) -> None:
             job_ids = await self.pool.zrangebyscore(
                 self.queue_name, offset=self._queue_read_offset, count=count, max=now
             )
+            self.abort_tasks = [
+                k[len(abort_key_prefix):]
```
Comment (samuelcolvin) [inline arq/worker.py]: `len(abort_key_prefix)` can be a module level constant.

File: arq/worker.py [hunk 10]
```diff
@@ -571,7 +592,7 @@ def handle_sig(self, signum):
             self.jobs_retried,
             len(self.tasks),
         )
-        for t in self.tasks:
+        for jid, t in self.tasks.items():
```
Comment (samuelcolvin) [inline arq/worker.py]:
```suggestion
        for t in self.tasks.values():
```
?

File: arq/worker.py [hunk 11]
```diff
@@ -580,7 +601,7 @@ def handle_sig(self, signum):
     async def close(self):
         if not self.pool:
             return
-        await asyncio.gather(*self.tasks)
+        await asyncio.gather(*list(self.tasks.values()))
```
Comment (samuelcolvin) [inline arq/worker.py]: again, I think `list()` can be removed.

File: arq/worker.py [hunk 12]
```diff
@@ -450,9 +460,15 @@ def run(self) -> None:
...
                 self.jobs_retried += 1
-            elif self.retry_jobs and isinstance(e, asyncio.CancelledError):
-                logger.info('%6.2fs ↻ %s cancelled, will be run again', t, ref)
-                self.jobs_retried += 1
+            elif isinstance(e, asyncio.CancelledError):
+                if job_id in self.abort_tasks:
+                    logger.info('%6.2fs 🛇  %s aborted', t, ref)
+                    result = e
+                    finish = True
+                    self.abort_tasks.remove(job_id)
```
Comment (samuelcolvin) [inline arq/worker.py]: we should delete the abort key here too.

File: arq/worker.py (hunk 7 above)
Comment (david-e) [reply arq/worker.py]: I used in a task like this:
```python
async def cronjob(ctx, funcname, crontab, run_at_startup=False, *args, **kwargs):
    rds = ctx['redis']
    cj = CronTab(crontab)  # python-crontab string parser
    while True:
        dt = datetime.now()
        nr = cj.next(dt, default_utc=True)
        if nr or run_at_startup:
            run_at_startup = False
            await asyncio.sleep(nr)
        await rds.enqueue_job(funcname, *args, **kwargs)
 ```

and then define crontab in this way:

```python
func(cronjob, timeout=0)
```
If I pass None as timeout I cannot have the desired behaviour because of this, in worker.py:
```python
timeout_s = self.job_timeout_s if function.timeout_s is None else function.timeout_s
```
So I can schedule crontab remotely, start and stop them when needed, without the limit of timeout.

File: arq/worker.py (hunk 5 above)
Comment (david-e) [reply arq/worker.py]: Sure, i changed a bit the code to use it more efficiently (I think). The results is something like that:

```python
    async def _abort_jobs(self):
        todel, cur = [], b'0'
        lk = len(abort_key_prefix)
        pool, tasks = self.pool, self.tasks
        while cur:
            cur, keys = await pool.scan(cur, match=f'{abort_key_prefix}*')
            for job_key in keys:
                job_id = job_key[lk:]
                if job_id not in tasks:
                    todel.append(job_key)
                    continue
                tasks[job_id].cancel()
                self._aborting_taks.add(job_id)
        if todel:
            await pool.delete(*todel)

    async def _poll_iteration(self):
        count = self.queue_read_limit
        if self.burst and self.max_burst_jobs >= 0:
            burst_jobs_remaining = self.max_burst_jobs - self._jobs_started()
            if burst_jobs_remaining < 1:
                return
            count = min(burst_jobs_remaining, count)

        async with self.sem:  # don't bother with zrangebyscore until we have "space" to run the jobs
            now = timestamp_ms()
            job_ids = await self.pool.zrangebyscore(
                self.queue_name, offset=self._queue_read_offset, count=count, max=now
            )
            if self.abort_jobs:
                await self._abort_jobs()
        await self.run_jobs(job_ids)

        for t in list(self.tasks.values()):
            if t.done():
                del self.tasks[job_id]
                # required to make sure errors in run_job get propagated
                t.result()

        await self.heart_beat()
```
What do you think?

File: arq/worker.py (hunk 6 above)
Comment (david-e) [reply arq/worker.py]: ok

File: arq/worker.py (hunk 8 above)
Comment (david-e) [reply arq/worker.py]: I fix it in this way:

```python
elif isinstance(e, asyncio.CancelledError) and job_id in self._aborting_tasks:
    logger.info('%6.2fs 🛇  %s aborted', t, ref)
    result = e
    finish = True
    self._aborting_tasks.remove(job_id)
   # redis key will be removed at next scan
elif isinstance(e, asyncio.CancelledError) and self.retry_jobs:
    logger.info('%6.2fs ↻ %s cancelled, will be run again', t, ref)
    self.jobs_retried += 1
```

Comment (samuelcolvin) [reply arq/worker.py]: I think this is less clear and duplicates the isinstance check.

**Comment by david-e** [comment]
I have one question: why do you use polling every 0.5s (or custom interval) and not the publish/subscribe mechaniscm provided from redis?
subsribing to

**Comment by samuelcolvin** [comment]
we could use publish and subscribe, but it would be a fair amount of work to setup.

The question is: why bother? Redis is amazingly performant, a few workers and clients polling it every 0.5s isn't going to have any effect on performance, even if reduced the timeout to 0.1s with 10 clients/workers, redis wouldn't notice in terms of throughput.

The only reason would be to reduce latency for jobs, responses. But if the worker is busy the polling delay won't have any effect as it'll make the request immediately when it's ready for more jobs.

In short, because I felt it wasn't worth. I'm prepared to be convinced otherwise, if you want to try, please create a new issue.

**Comment by david-e** [comment]
> we could use publish and subscribe, but it would be a fair amount of work to setup.
> 
> The question is: why bother? Redis is amazingly performant, a few workers and clients polling it every 0.5s isn't going to have any effect on performance, even if reduced the timeout to 0.1s with 10 clients/workers, redis wouldn't notice in terms of throughput.
> 
> The only reason would be to reduce latency for jobs, responses. But if the worker is busy the polling delay won't have any effect as it'll make the request immediately when it's ready for more jobs.
> 
> In short, because I felt it wasn't worth. I'm prepared to be convinced otherwise, if you want to try, please create a new issue.

You are probably right. Publish/subscribe is the way I probably start with for a project like this, but these code works and it is really good, so I agree with you that the effort is a waste of time. I like the way you organize the project and the clearness of code. I hope to be helpful like you are for me.

File: arq/connections.py [hunk 13]
```diff
@@ -101,6 +108,8 @@ async def enqueue_job(
         expires_ms = to_ms(_expires)
 
         with await self as conn:
+            if _del_prev_result:
+                await conn.delete(result_key_prefix + job_id)
```
Comment (samuelcolvin) [inline arq/connections.py:112]: why can't this be part of the pipelined operations?

Comment (samuelcolvin) [reply arq/connections.py:112]: why is this needed as part of this PR.

I think you may need to separate your work here into multiple pull requests for easier reveiwing.

File: arq/connections.py [hunk 14]
```diff
@@ -133,34 +144,58 @@ async def enqueue_job(
                 return
         return Job(job_id, redis=self, _queue_name=_queue_name, _deserializer=self.job_deserializer)
 
-    async def _get_job_result(self, key) -> JobResult:
-        job_id = key[len(result_key_prefix) :]
+    async def _get_job_result(self, key, partial: bool = False) -> JobResult:
```
Comment (samuelcolvin) [inline arq/connections.py:147]: please can you explain the idea behind "partial" results - I think it's added since I last reviewed this PR?

Again, I think this might need to be in a separate PR.

File: arq/connections.py [hunk 15]
```diff
@@ -133,34 +144,58 @@ async def enqueue_job(
...
+        return Job(job_id=job_id, redis=self, _queue_name=queue_name, _deserializer=self.job_deserializer)
+
     async def queued_jobs(self, *, queue_name: str = default_queue_name) -> List[JobDef]:
         """
         Get information about queued, mostly useful when testing.
         """
         jobs = await self.zrange(queue_name, withscores=True)
         return await asyncio.gather(*[self._get_job_def(job_id, score) for job_id, score in jobs])
 
+    async def all_jobs_queued(self, *, queue_name: str = default_queue_name) -> List[JobDef]:
```
Comment (samuelcolvin) [inline arq/connections.py:184]: why is this method required?

Again I don't think it's related to this PR.

File: arq/connections.py [hunk 16]
```diff
@@ -133,34 +144,58 @@ async def enqueue_job(
...
+        """
+        Get information about queued jobs, same of queued_jobs, but adopting naming convention.
+        """
+        return await self.queued_jobs(queue_name=queue_name)
+
+    async def all_jobs_running(self) -> List[JobDef]:
```
Comment (samuelcolvin) [inline arq/connections.py:190]: same

File: arq/connections.py [hunk 17]
```diff
@@ -133,34 +144,58 @@ async def enqueue_job(
...
+    async def _get_job_result(self, key, partial: bool = False) -> JobResult:
+        job_id = key[len(result_key_prefix if not partial else partial_result_key_prefix) :]
         job = Job(job_id, self, _deserializer=self.job_deserializer)
-        r = await job.result_info()
+        r = await job.result_info(partial)
         r.job_id = job_id
         return r
 
-    async def all_job_results(self) -> List[JobResult]:
+    async def all_jobs_results(self) -> List[JobResult]:
```
[the remark this answers is not in the document]
Comment (david-e) [reply arq/connections.py]: Sorry, I continue to develop my changes on the same branch. Now I revert it for the only purpose of this PR, moving other stuff in a personal branch.

File: arq/jobs.py [hunk 18]
```diff
@@ -66,6 +66,15 @@ def __init__(
         self._queue_name = _queue_name
         self._deserializer = _deserializer
 
+    async def abort(
+        self, timeout: Optional[float] = None, *, pole_delay: float = 0.5, key_expire: Optional[int] = 10
+    ) -> Any:
```
Comment (samuelcolvin) [inline arq/jobs.py:71]: I'm not clear on the return type here - what does it mean? I think it should be `bool`

File: arq/worker.py (hunk 7 above)
Comment (samuelcolvin) [reply arq/worker.py:459]: please remove this.

File: docs/index.rst [hunk 19]
```diff
@@ -167,6 +167,10 @@ optionally with a duration to defer rerunning the jobs by:
 
 .. literalinclude:: examples/retry.py
 
+It is also possible to abort long running tasks, passing ``abort_jobs=True`` to ``WorkerSettins``.
```
Comment (samuelcolvin) [inline docs/index.rst:170]:
```suggestion
It is also possible to abort long running tasks, passing ``abort_jobs=True`` to ``WorkerSettings``.
```

File: arq/worker.py [hunk 20]
```diff
@@ -275,13 +279,27 @@ async def main(self):
...
+        todel, cur = [], b'0'
+        lk, mk = len(abort_key_prefix), f'{abort_key_prefix}*'
+        while cur:
+            cur, keys = await self.pool.scan(cur, match=mk)
+            for job_key in keys:
+                job_id = job_key[lk:]
+                if job_id not in self.tasks:
+                    todel.append(job_key)
+                    continue
+                self._aborting_tasks.add(job_id)
```
Comment (samuelcolvin) [inline arq/worker.py:299]:
```suggestion
                if job_id in self.tasks:
                    self._aborting_tasks.add(job_id)
                else:
                    todel.append(job_key)
```
slightly cleaner.

**Comment by samuelcolvin** [comment]
Closing this as inactive, but I can re-open or review a new PR if a fix becomes available.
